src/ILCompiler.ReadyToRun/src/IBC/ibcdatareader.cs

```diff
@@ -0,0 +1,724 @@
+
+            minorVersion = 0;
+
+            if (remaining >= sizeof(uint))
+            {
+                minorVersion = reader.ReadUInt32();
+                remaining -= sizeof(int);
+            }
+
+            minified = false;
+            partialNGen = false;
+
+            if (remaining >= sizeof(uint))
+            {
+                var fileFlags = (IBC.Constants.FileFlags)reader.ReadUInt32();
+                minified = fileFlags.HasFlag(IBC.Constants.FileFlags.Minified);
+                partialNGen = fileFlags.HasFlag(IBC.Constants.FileFlags.PartialNGen);
+
+                if (minified && (majorVersion < IBC.Constants.LowestMajorVersionSupportingMinify))
+                {
+                    throw new IBCException("Minified flag is set but the file's version should not support it.");
+                }
+                remaining -= sizeof(uint);
+            }
+
+            SeekTo(startPosition + size);
+
+            // Make sure that CompatibleVersion <= ReadVersion <= CurrentVersion
+            // (Ngen only pays attention to major version, but ideally even a new
+            // minor version should be accompanied by a new version of IBCMerge.)
+
+            if ((majorVersion > IBC.Constants.CurrentMajorVersion) ||
+                ((majorVersion == IBC.Constants.CurrentMajorVersion) && (minorVersion > IBC.Constants.CurrentMinorVersion)) ||
+                (majorVersion < IBC.Constants.CompatibleMajorVersion) ||
+                ((majorVersion == IBC.Constants.CompatibleMajorVersion) && (minorVersion < IBC.Constants.CompatibleMinorVersion)))
+            {
+                throw new IBCException(String.Format("Version mismatch: {0}.{1} not between {2}.{3} and {4}.{5}",
+                    majorVersion,
+                    minorVersion,
+                    IBC.Constants.CompatibleMajorVersion,
+                    IBC.Constants.CompatibleMinorVersion,
+                    IBC.Constants.CurrentMajorVersion,
+                    IBC.Constants.CurrentMinorVersion));
+            }
+        }
+
+        void ReadFooter()
+        {
+            uint endToken = reader.ReadUInt32();
+            if (endToken != IBC.Constants.FooterMagicNumber)
+            {
+                throw new IBCException("Invalid file footer");
+            }
+        }
+
+        private struct SectionInfo
+        {
+            public long Offset;
+            public long Length;
+        }
+
+        Dictionary<SectionFormat, SectionInfo> ReadSectionTable(out long highestValidOffset)
+        {
+            uint NumberOfEntries = reader.ReadUInt32();
+            var result = new Dictionary<SectionFormat, SectionInfo>();
+
+            highestValidOffset = 0;
+            for (int i = 0; i < NumberOfEntries; ++i)
+            {
+                SectionInfo sectionInfo;
+                SectionFormat sectionFormat = (SectionFormat)reader.ReadInt32();
+
+                if ((majorVersion == 1) &&
```

Comment:
If we're creating our own clone of this code, it might be worth a bit of cleanup w.r.t. older IBC versions we no longer intend to support (as indicated by the MajorVersion == 1 exception throw above).

Reply:
At this time, I'm trying to keep this file very similar to the closed source version. We may branch out soon, but I'm not ready to do so yet.